File: src/main/java/com/lp/v2/domains/auth/AuthController.java
package com.lp.v2.domains.auth;

import com.lp.v2.domains.auth.dto.AuthLogInReq;
import com.lp.v2.domains.auth.dto.AccessTokenRes;
import com.lp.v2.domains.auth.dto.TokenPair;
import com.lp.v2.domains.auth.service.AuthService;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.Duration;

@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;

    @PostMapping("/login")
    public ResponseEntity<AccessTokenRes> logIn (@RequestBody AuthLogInReq req) {
        TokenPair tokenPair = authService.logIn(req);

        ResponseCookie cookie = ResponseCookie.from("refreshToken", tokenPair.refreshToken())
                .httpOnly(true)
                .secure(true)
                .path("/")
                .maxAge(604800000)
                .sameSite("Lax")
                .build();

        return ResponseEntity.ok()
                .header(HttpHeaders.SET_COOKIE, cookie.toString())
                .body(AccessTokenRes.of(tokenPair.accessToken()));
    }

    @PostMapping("/refresh")
    public ResponseEntity<AccessTokenRes> refresh(HttpServletRequest request) {
        String refreshToken = extractRefreshTokenFromCookie(request);
        if (refreshToken == null) {
            return ResponseEntity.badRequest().build();
        }

        TokenPair tokenPair = authService.reissueTokens(refreshToken);

        // 새로운 refresh token으로 쿠키 업데이트
        ResponseCookie cookie = ResponseCookie.from("refreshToken", tokenPair.refreshToken())
                .httpOnly(true)
                .secure(true)
                .path("/")
                .maxAge(Duration.ofDays(7).getSeconds())
                .sameSite("Lax")
                .build();

        return ResponseEntity.ok()
                .header(HttpHeaders.SET_COOKIE, cookie.toString())
                .body(AccessTokenRes.of(tokenPair.accessToken()));
    }

    private String extractRefreshTokenFromCookie(HttpServletRequest request) {
        if (request.getCookies() != null) {
            for (Cookie cookie : request.getCookies()) {
                if ("refreshToken".equals(cookie.getName())) {
                    return cookie.getValue();
                }
            }
        }
        return null;
    }


}

File: lp-api/src/main/kotlin/com/boaz/lp/api/auth/controller/AuthController.kt
package com.boaz.lp.api.auth.controller

import com.boaz.lp.api.auth.service.AuthService
import com.boaz.lp.common.dto.ApiResponse
import com.boaz.lp.common.exception.BusinessException
import com.boaz.lp.common.exception.ErrorCode
import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletResponse
import org.springframework.web.bind.annotation.CookieValue
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/auth")
class AuthController(
    private val authService: AuthService
) {

    @PostMapping("/refresh")
    fun refresh(
        @CookieValue("refresh_token", required = false) refreshToken: String?,
        response: HttpServletResponse
    ): ApiResponse<Map<String, String>> {
        if (refreshToken == null) throw BusinessException(ErrorCode.TOKEN_NOT_FOUND)
        val accessToken = authService.refresh(refreshToken)
        return ApiResponse.success(mapOf("accessToken" to accessToken))
    }

    @PostMapping("/logout")
    fun logout(
        @CookieValue("refresh_token", required = false) refreshToken: String?,
        response: HttpServletResponse
    ): ApiResponse<Unit> {
        authService.logout(refreshToken)
        response.addCookie(Cookie("refresh_token", "").apply {
            isHttpOnly = true
            secure = true
            path = "/auth/refresh"
            maxAge = 0
        })
        return ApiResponse.success(Unit)
    }
}
